{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os.path\n",
    "import numpy as np\n",
    "import PIL\n",
    "import glob\n",
    "from PIL import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>statistic</th>\n",
       "      <th>lungc</th>\n",
       "      <th>hdac</th>\n",
       "      <th>who</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pearson</td>\n",
       "      <td>lungc_kpc</td>\n",
       "      <td>hdac_kpc</td>\n",
       "      <td>WHO_kpc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>spearman</td>\n",
       "      <td>lungc_ksc</td>\n",
       "      <td>hdac_ksc</td>\n",
       "      <td>WHO_ksc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>kendall</td>\n",
       "      <td>lungc_kkc</td>\n",
       "      <td>hdac_kkc</td>\n",
       "      <td>WHO_kkc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mine</td>\n",
       "      <td>lungc_mine</td>\n",
       "      <td>hdac_mine</td>\n",
       "      <td>WHO_mine</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  statistic       lungc       hdac       who\n",
       "0   pearson   lungc_kpc   hdac_kpc   WHO_kpc\n",
       "1  spearman   lungc_ksc   hdac_ksc   WHO_ksc\n",
       "2   kendall   lungc_kkc   hdac_kkc   WHO_kkc\n",
       "3      mine  lungc_mine  hdac_mine  WHO_mine"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/Users/KevinBu/Desktop/clemente_lab/CUTIE/')\n",
    "stat_df = pd.read_csv('scripts/stat_dict_real.txt', sep = '\\t')\n",
    "rev_stat_df = pd.read_csv('scripts/rev_stat_dict_real.txt', sep = '\\t')\n",
    "stat_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "headers = [\n",
    "    'analysis_id',\n",
    "    'distribution',\n",
    "    'statistic',\n",
    "    'defaulted', # binary\n",
    "    'initial_corr',\n",
    "    'true_corr(TP_FN)',\n",
    "    'false_corr(FP_TN)',\n",
    "    'rs_true_corr_TP_FN',\n",
    "    'rs_false_corr_FP_TN',\n",
    "    'runtime'\n",
    "]\n",
    "\n",
    "    \n",
    "def parse_log(f):\n",
    "    lines = [l.strip() for l in f.readlines()]\n",
    "    defaulted = False\n",
    "    # check if FDR correction defaulted\n",
    "    for l in lines:\n",
    "        if \"defaulted\" in l: \n",
    "            defaulted = True\n",
    "        elif \"initial_corr\" in l:\n",
    "            initial_corr = float(l.split(' ')[-1])\n",
    "        elif \"false correlations\" in l:\n",
    "            false_corr = float(l.split(' ')[-1])\n",
    "        elif \"true correlations\" in l:\n",
    "            true_corr = float(l.split(' ')[-1])\n",
    "        elif \"FP/TN1\" in l:\n",
    "            rs_false = float(l.split(' ')[-1])\n",
    "        elif \"TP/FN1\" in l:\n",
    "            rs_true = float(l.split(' ')[-1])\n",
    "        elif \"runtime\" in l:\n",
    "            runtime = float(l.split(' ')[-1])\n",
    "    return defaulted, initial_corr, false_corr, true_corr, rs_false, rs_true, runtime\n",
    "\n",
    "\n",
    "m_to_stat = {\n",
    "    'kpc': 'pearson',\n",
    "    'ksc': 'spearman',\n",
    "    'kkc': 'kendall',\n",
    "    'mine': 'mine'        \n",
    "}\n",
    "for_df = pd.DataFrame()\n",
    "\n",
    "for dataset in stat_df.columns.values[1:4]:\n",
    "    # statistic = 'pearson'\n",
    "    subset_df = stat_df.set_index('statistic').loc[:,dataset]\n",
    "    rev_subset_df = rev_stat_df.set_index('statistic').loc[:,dataset]\n",
    "        \n",
    "    # for each distribution\n",
    "#    for dist in distributions:\n",
    " #       # dist = 'sim_copula_n50_norm_0_1'\n",
    "        \n",
    "    for m in subset_df:\n",
    "        # m = 'kpc'\n",
    "        path = '../Submissions/CUtIe/final_data_fixed/cutie_' + m + '1fdr0.05/'\n",
    "        analysis_id = str('cutie_' + m + '1fdr0.05')\n",
    "        files = sorted(glob.glob(path + '*.txt'))\n",
    "        # grab most recent log file\n",
    "        rel_logfile = files[-1]\n",
    "        with open(rel_logfile, 'r') as f:\n",
    "            try: \n",
    "                defaulted, initial_corr, false_corr, true_corr, rs_false, rs_true, runtime = parse_log(f)\n",
    "\n",
    "                new_row = pd.DataFrame([[analysis_id, dataset, m_to_stat[m.split('_')[-1]], \n",
    "                                        defaulted, initial_corr, true_corr, false_corr, rs_true, rs_false, runtime]], columns=headers)\n",
    "\n",
    "                for_df = for_df.append(new_row)\n",
    "            except: \n",
    "                print(analysis_id)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>analysis_id</th>\n",
       "      <th>distribution</th>\n",
       "      <th>statistic</th>\n",
       "      <th>defaulted</th>\n",
       "      <th>initial_corr</th>\n",
       "      <th>true_corr(TP_FN)</th>\n",
       "      <th>false_corr(FP_TN)</th>\n",
       "      <th>rs_true_corr_TP_FN</th>\n",
       "      <th>rs_false_corr_FP_TN</th>\n",
       "      <th>runtime</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_kpc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>pearson</td>\n",
       "      <td>False</td>\n",
       "      <td>15458.0</td>\n",
       "      <td>1098.0</td>\n",
       "      <td>14360.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11396.0</td>\n",
       "      <td>1079.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_ksc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>spearman</td>\n",
       "      <td>False</td>\n",
       "      <td>28304.0</td>\n",
       "      <td>9092.0</td>\n",
       "      <td>19212.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13542.0</td>\n",
       "      <td>3723.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_kkc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>kendall</td>\n",
       "      <td>False</td>\n",
       "      <td>27636.0</td>\n",
       "      <td>8796.0</td>\n",
       "      <td>18840.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13240.0</td>\n",
       "      <td>2245.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_mine1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>mine</td>\n",
       "      <td>False</td>\n",
       "      <td>580.0</td>\n",
       "      <td>496.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>676.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_hdac_kpc1fdr0.05</td>\n",
       "      <td>hdac</td>\n",
       "      <td>pearson</td>\n",
       "      <td>False</td>\n",
       "      <td>4122.0</td>\n",
       "      <td>3470.0</td>\n",
       "      <td>652.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>599.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                analysis_id distribution statistic  defaulted  initial_corr  \\\n",
       "0   cutie_lungc_kpc1fdr0.05        lungc   pearson      False       15458.0   \n",
       "0   cutie_lungc_ksc1fdr0.05        lungc  spearman      False       28304.0   \n",
       "0   cutie_lungc_kkc1fdr0.05        lungc   kendall      False       27636.0   \n",
       "0  cutie_lungc_mine1fdr0.05        lungc      mine      False         580.0   \n",
       "0    cutie_hdac_kpc1fdr0.05         hdac   pearson      False        4122.0   \n",
       "\n",
       "   true_corr(TP_FN)  false_corr(FP_TN)  rs_true_corr_TP_FN  \\\n",
       "0            1098.0            14360.0                 0.0   \n",
       "0            9092.0            19212.0                 0.0   \n",
       "0            8796.0            18840.0                 0.0   \n",
       "0             496.0               84.0                 0.0   \n",
       "0            3470.0              652.0                 0.0   \n",
       "\n",
       "   rs_false_corr_FP_TN  runtime  \n",
       "0              11396.0  1079.91  \n",
       "0              13542.0  3723.45  \n",
       "0              13240.0  2245.40  \n",
       "0                  0.0   676.08  \n",
       "0                  0.0   599.00  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "for_df.to_csv('../Submissions/CUtIe/final_data_fixed/final_real_df.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = [\n",
    "    'Microbiome',\n",
    "    'Gene Expression',\n",
    "    'WHO'\n",
    "]\n",
    "\n",
    "index = [\n",
    "    'Pearson',\n",
    "    '% initial sig',\n",
    "    'Spearman',\n",
    "    '% initial sig',\n",
    "    'Kendall',\n",
    "    '% initial sig',\n",
    "    'MIC',\n",
    "    '% initial sig',\n",
    "\n",
    "]\n",
    "dists = ['lungc', 'hdac', 'who'] \n",
    "method = ['cutie', 'jackknife', 'bootstrap', 'cookd']\n",
    "dist_to_corr = {\n",
    "    'lungc': 748 * 747,\n",
    "    'hdac': 100 * 100,\n",
    "    'who': 354 * 353\n",
    "}\n",
    "\n",
    "df_array = []\n",
    "for stat in stat_df['statistic']:\n",
    "    # stat = 'Pearson'\n",
    "    row_fracs = []\n",
    "    for dist in dists:\n",
    "        row = for_df[(for_df['distribution'] == dist) & (for_df['statistic'] == stat)] \n",
    "        row_fracs.append(float(row['true_corr(TP_FN)'] /row['initial_corr'].values)) # correctly id tp\n",
    "    df_array.append(row_fracs)\n",
    "    \n",
    "    initial_sig_fracs = []\n",
    "    for dist in dists:\n",
    "        row = for_df[(for_df['distribution'] == dist) & (for_df['statistic'] == stat)] \n",
    "        # change number 249500 to n_corr depending on dataset\n",
    "        initial_sig_fracs.append(float(row['initial_corr'] / dist_to_corr[dist]))\n",
    "    df_array.append(initial_sig_fracs)\n",
    "    \n",
    "new_df = pd.DataFrame(data = df_array, index = index, columns = header) \n",
    "new_df = new_df.rename_axis('Statistic')\n",
    "#new_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df.to_csv('../Submissions/CUtIe/final_data_fixed/real_tpfp.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "headers = [\n",
    "    'analysis_id',\n",
    "    'distribution',\n",
    "    'statistic',\n",
    "    'defaulted', # binary\n",
    "    'initial_corr',\n",
    "    'true_corr(TP_FN)',\n",
    "    'false_corr(FP_TN)',\n",
    "    'rs_true_corr_TP_FN',\n",
    "    'rs_false_corr_FP_TN',\n",
    "    'runtime'\n",
    "]\n",
    "\n",
    "    \n",
    "def parse_log(f):\n",
    "    lines = [l.strip() for l in f.readlines()]\n",
    "    defaulted = False\n",
    "    # check if FDR correction defaulted\n",
    "    for l in lines:\n",
    "        if \"defaulted\" in l: \n",
    "            defaulted = True\n",
    "        elif \"initial_corr\" in l:\n",
    "            initial_corr = float(l.split(' ')[-1])\n",
    "        elif \"false correlations\" in l:\n",
    "            false_corr = float(l.split(' ')[-1])\n",
    "        elif \"true correlations\" in l:\n",
    "            true_corr = float(l.split(' ')[-1])\n",
    "        elif \"FP/TN1\" in l:\n",
    "            rs_false = float(l.split(' ')[-1])\n",
    "        elif \"TP/FN1\" in l:\n",
    "            rs_true = float(l.split(' ')[-1])\n",
    "        elif \"runtime\" in l:\n",
    "            runtime = float(l.split(' ')[-1])\n",
    "            \n",
    "    return defaulted, initial_corr, false_corr, true_corr, rs_false, rs_true, runtime\n",
    "\n",
    "\n",
    "m_to_stat = {\n",
    "    'rpc': 'pearson',\n",
    "    'rsc': 'spearman',\n",
    "    'rkc': 'kendall',\n",
    "    'rmine': 'mine'        \n",
    "}\n",
    "rev_df = pd.DataFrame()\n",
    "\n",
    "for dataset in rev_stat_df.columns.values[1:4]:\n",
    "    # statistic = 'pearson'\n",
    "    subset_df = rev_stat_df.set_index('statistic').loc[:,dataset]\n",
    "    rev_subset_df = rev_stat_df.set_index('statistic').loc[:,dataset]\n",
    "        \n",
    "    # for each distribution\n",
    "#    for dist in distributions:\n",
    " #       # dist = 'sim_copula_n50_norm_0_1'\n",
    "        \n",
    "    for m in rev_subset_df:\n",
    "        # m = 'kpc'\n",
    "        path = '../Submissions/CUtIe/final_data_fixed/cutie_' + m + '1fdr0.05/'\n",
    "        analysis_id = str('cutie_' + m + '1fdr0.05')\n",
    "        files = sorted(glob.glob(path + '*.txt'))\n",
    "        # grab most recent log file\n",
    "        rel_logfile = files[-1]\n",
    "        with open(rel_logfile, 'r') as f:\n",
    "            try: \n",
    "                defaulted, initial_corr, false_corr, true_corr, rs_false, rs_true, runtime = parse_log(f)\n",
    "\n",
    "                new_row = pd.DataFrame([[analysis_id, dataset, m_to_stat[m.split('_')[-1]], \n",
    "                                        defaulted, initial_corr, true_corr, false_corr, rs_true, rs_false, runtime]], columns=headers)\n",
    "\n",
    "                rev_df = rev_df.append(new_row)\n",
    "            except: \n",
    "                print(analysis_id)\n",
    "                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>analysis_id</th>\n",
       "      <th>distribution</th>\n",
       "      <th>statistic</th>\n",
       "      <th>defaulted</th>\n",
       "      <th>initial_corr</th>\n",
       "      <th>true_corr(TP_FN)</th>\n",
       "      <th>false_corr(FP_TN)</th>\n",
       "      <th>rs_true_corr_TP_FN</th>\n",
       "      <th>rs_false_corr_FP_TN</th>\n",
       "      <th>runtime</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_rpc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>pearson</td>\n",
       "      <td>False</td>\n",
       "      <td>543298.0</td>\n",
       "      <td>13260.0</td>\n",
       "      <td>530038.0</td>\n",
       "      <td>297502.0</td>\n",
       "      <td>10142.0</td>\n",
       "      <td>31306.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_rsc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>spearman</td>\n",
       "      <td>False</td>\n",
       "      <td>530452.0</td>\n",
       "      <td>9524.0</td>\n",
       "      <td>520928.0</td>\n",
       "      <td>318706.0</td>\n",
       "      <td>4970.0</td>\n",
       "      <td>61934.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_rkc1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>kendall</td>\n",
       "      <td>False</td>\n",
       "      <td>531120.0</td>\n",
       "      <td>9314.0</td>\n",
       "      <td>521806.0</td>\n",
       "      <td>319334.0</td>\n",
       "      <td>4866.0</td>\n",
       "      <td>38631.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_lungc_rmine1fdr0.05</td>\n",
       "      <td>lungc</td>\n",
       "      <td>mine</td>\n",
       "      <td>False</td>\n",
       "      <td>558176.0</td>\n",
       "      <td>338.0</td>\n",
       "      <td>557838.0</td>\n",
       "      <td>252450.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>51888.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cutie_hdac_rpc1fdr0.05</td>\n",
       "      <td>hdac</td>\n",
       "      <td>pearson</td>\n",
       "      <td>False</td>\n",
       "      <td>5778.0</td>\n",
       "      <td>768.0</td>\n",
       "      <td>5010.0</td>\n",
       "      <td>1306.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>853.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 analysis_id distribution statistic  defaulted  initial_corr  \\\n",
       "0    cutie_lungc_rpc1fdr0.05        lungc   pearson      False      543298.0   \n",
       "0    cutie_lungc_rsc1fdr0.05        lungc  spearman      False      530452.0   \n",
       "0    cutie_lungc_rkc1fdr0.05        lungc   kendall      False      531120.0   \n",
       "0  cutie_lungc_rmine1fdr0.05        lungc      mine      False      558176.0   \n",
       "0     cutie_hdac_rpc1fdr0.05         hdac   pearson      False        5778.0   \n",
       "\n",
       "   true_corr(TP_FN)  false_corr(FP_TN)  rs_true_corr_TP_FN  \\\n",
       "0           13260.0           530038.0            297502.0   \n",
       "0            9524.0           520928.0            318706.0   \n",
       "0            9314.0           521806.0            319334.0   \n",
       "0             338.0           557838.0            252450.0   \n",
       "0             768.0             5010.0              1306.0   \n",
       "\n",
       "   rs_false_corr_FP_TN   runtime  \n",
       "0              10142.0  31306.66  \n",
       "0               4970.0  61934.97  \n",
       "0               4866.0  38631.15  \n",
       "0                  0.0  51888.13  \n",
       "0                  6.0    853.05  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rev_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "rev_df.to_csv('../Submissions/CUtIe/final_data_fixed/final_real_df_rev.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = [\n",
    "    'Microbiome',\n",
    "    'Gene Expression',\n",
    "    'WHO'\n",
    "]\n",
    "\n",
    "index = [\n",
    "    'Pearson',\n",
    "    '% initial insig',\n",
    "    'Spearman',\n",
    "    '% initial insig',\n",
    "    'Kendall',\n",
    "    '% initial insig',\n",
    "    'MIC',\n",
    "    '% initial insig',\n",
    "\n",
    "]\n",
    "dists = ['lungc', 'hdac', 'who'] \n",
    "method = ['cutie', 'jackknife', 'bootstrap', 'cookd']\n",
    "dist_to_corr = {\n",
    "    'lungc': 748 * 747,\n",
    "    'hdac': 100 * 100,\n",
    "    'who': 354 * 353\n",
    "}\n",
    "\n",
    "df_array = []\n",
    "for stat in rev_stat_df['statistic']:\n",
    "    # stat = 'Pearson'\n",
    "    row_fracs = []\n",
    "    for dist in dists:\n",
    "        row = rev_df[(rev_df['distribution'] == dist) & (rev_df['statistic'] == stat)] \n",
    "        row_fracs.append(float(row['true_corr(TP_FN)'] /row['initial_corr'].values)) # correctly id tp\n",
    "    df_array.append(row_fracs)\n",
    "    \n",
    "    initial_sig_fracs = []\n",
    "    for dist in dists:\n",
    "        row = rev_df[(rev_df['distribution'] == dist) & (rev_df['statistic'] == stat)] \n",
    "        # change number 249500 to n_corr depending on dataset\n",
    "        initial_sig_fracs.append(float(row['initial_corr'] / dist_to_corr[dist]))\n",
    "    df_array.append(initial_sig_fracs)\n",
    "    \n",
    "new_df = pd.DataFrame(data = df_array, index = index, columns = header) \n",
    "new_df = new_df.rename_axis('Statistic')\n",
    "#new_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df.to_csv('../Submissions/CUtIe/final_data_fixed/real_tnfn.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
